tf_trees/tel.py

```diff
@@ -0,0 +1,190 @@
+@ops.RegisterGradient('NTComputeOutputOp')
+def _nt_compute_input_and_internal_params_gradients_op_cc(
+    op, grad_loss_wrt_tree_output, _):
+  """Associated a custom gradient with an op."""
+  output_logits_dim = op.get_attr('output_logits_dim')
+  depth = op.get_attr('depth')
+  parallelize_over_samples = op.get_attr('parallelize_over_samples')
+  return [tf_trees_module.nt_compute_input_and_internal_params_gradients_op(
+      grad_loss_wrt_tree_output, op.inputs[0], op.inputs[1], op.inputs[2],
+      op.inputs[3], output_logits_dim, depth, parallelize_over_samples), None]
+
+
+class TEL(keras.layers.Layer):
+    """A custom layer containing additive decision trees.
+
+    Each tree in the layer is composed of splitting (internal) nodes and leaves.
+    A splitting node "routes" the samples left or right based on the
+    corresponding activation. Samples can be routed in a hard way (i.e., sent
+    to only one child) or in a soft way. The decision whether to hard or soft
+    route is controlled by the smooth_step_param (see details below).
+    The trees are modeled using smooth functions and can be optimized
+    using standard continuous optimization methods (e.g., SGD).
+
+    The layer can be combined with other Keras layers and can be used anywhere
+    in the neural net.
+
+    Attributes:
+      output_logits_dim: Dimension of the output.
+      trees_num: Number of trees in the layer.
+      depth: Depth of each tree.
+      smooth_step_param: A non-negative float. Larger values make the trees
+        more likely to hard route samples (i.e., samples reach fewer leaves).
+        Note smooth_step_param = 1/gamma, where gamma is the parameter defined
+        in the TEL paper. 
+      sum_outputs: Boolean. If true, the outputs of the trees will be added,
+        leading to a 2D tensor of shape=[batch_size, output_logits_dim].
+        Otherwise, the tree outputs are not added and the layer output is
+        a 2D tensor of shape=[batch_size, trees_num * output_logits_dim].
+      parallelize_over_samples: Boolean, If true, parallelizes the updates over
+        the samples in each tree. Might lead to speedups when the number of
+        trees is small (at the cost of extra memory consumption).
+      split_initializer: A Keras initializer for the internal (splitting) nodes.
+      leaf_initializer: A Keras initializer for the leaves.
+      split_regularizer: A Keras regularizer for the internal (splitting) nodes.
+      leaf_regularizer: A Keras regularizer for the leaves.
+    Input shape: A tensor of shape=[batch_size, input_dim].
+    Output shape: A tensor of shape=[batch_size, output_logits_dim] if
+      sum_outputs=True. Otherwise, a tensor of shape=[batch_size, trees_num *
+      output_logits_dim].
+    """
+
+    def __init__(self,
+                 output_logits_dim,
+                 trees_num=1,
+                 depth=3,
+                 smooth_step_param=1.0,
+                 sum_outputs=True,
```

Comment:
was not it performing dense computation with value of 1?

Reply:
you mean for the smooth_step_param? 1 is fast in our experiments

Reply:
oh it is gamma i think i was confusing it with. 
But anyway - should we perform some validation on this param to make sure we won't do dense computation. Eg what is the reasonable range